### DISCUSSION TITLE
custom bootstrapper modification with the same feature selection doesn't rerun MSI package

### DISCUSSION THREAD

**Discussion by demidov250471** [wixtoolset/issues#6607] [Questions]
Thank you, Rob, for your reply, but you've closed the issue prematurely.
I've tried "Repair" action long before posting the issue, both, planning it and/or selecting it in "PlanPackageBegin" event.
"Repair" mode ignores cmd properties, apparently, including Burn's <MsiProperty> as well. So it can't be communicated from the bootstrapper to the package.
On the other hand, repairing is not right, when I only need a "trivial" modification, without optimization. Bare MSI package does that, why doesn't the great bootstrapper? Any modification, whether features have changed or not, should work the same!

**Comment by barnson** [comment]
WiX doesn't use issues for support. We moved this issue to the [experimental GitHub discussions area][gh]. You might also consider asking your question on the [wix-users mailing list][ml] or posting on [Stack Overflow][so].

[gh]: https://github.com/wixtoolset/issues/discussions?discussions_q=category%3AQuestions
[ml]: http://wixtoolset.org/documentation/mailinglist/
[so]: https://stackoverflow.com/questions/tagged/wix

**Comment by barnson** [comment]
Burn doesn't run a package unless there's some action for it. As you already use features, set one of those features to be reinstalled and Burn will run it.

**Reply by demidov250471** [reply]
How? "PlanMsiFeature" event has an out parameter "FeatureState", which can't be "Reinstall".
And how can Burn decide, whether an action is needed, based only on feature selection? There can be properties as well!

**Reply by barnson** [reply]
When you set BOOTSTRAPPER_FEATURE_STATE_SOURCE when repair, Burn reinstalls the feature.

**Reply by demidov250471** [reply]
You probably mean local, not source...
As I've stated, I can't repair because it ignores command-line properties (and, thus, MsiProperty), so in "repair" mode you can't communicate properties from Burn to MSI.
I can live with this limitation for now. You can close the topic. Thank you.

**Reply by rseanhall** [reply]
If you look at [CalculateFeatureAction](https://github.com/wixtoolset/wix3/blob/b496a8fdc2979c9a8e70edd3a47f2e98fc178e9e/src/burn/engine/msiengine.cpp#L1532), both `LOCAL` and `SOURCE` would get it to be reinstalled.

Burn should send MsiProperty values during repair, if not then that's a bug. Windows Installer should not ignore command-line properties, are you talking about your MSI that ignores them during repair?

**Reply by demidov250471** [reply]
My MSI doesn't know what repair is, so it can't do something special during repair in particular.
As for repair mode per se, it seems to be Windows Installer limitation. See this thread:
https://www.advancedinstaller.com/forums/viewtopic.php?t=23683.
Actually, "repair" is not "modify", so it makes sense to ignore any intervention into excute sequence.
When I plan repair, all my properties start at their defaults.
On the other hand, repair is not right here, because I do modify the configuration - not via features, but via properties! And any modification should work the same, as it does in bare MSI, when I choose "Change" there and do not modify selection state of any feature - properties still get updated!

**Reply by rseanhall** [reply]
That link appears to point to an msiexec.exe limitation. Burn calls [`MsiInstallProduct`](https://docs.microsoft.com/en-us/windows/win32/api/msi/nf-msi-msiinstallproductw) directly for everything but uninstall (no difference at that level between install, modify, and repair) so properties should work during repair. As Bob said, Burn by design does not run a package unless there's some action for it. Repair is the only way that Burn supports today to force a package to run.

**Reply by demidov250471** [reply]
Well, they don't. What I do is a standard remeber-me pattern:
<img width="845" alt="Capture" src="https://user-images.githubusercontent.com/45342263/138420745-96f1e772-b843-4fbb-ac4c-2f8e24cb19b3.PNG">
This works when I run the MSI package, not the bootstrapper, choose "Change" and modify only the textual field linked to WEBAPIPORT.
In the bootstrapper I have:
...  MsiProperty Name="WEBAPIPORT" Value="[WebInterfacePort]" ...
and the corresponding WPF textual field. When bootstrapper is repaired, the property stays the same as it was found in the registry. But when it's modified with some other feature state changed, everything's right.

**Reply by rseanhall** [reply]
> When bootstrapper is repaired, the property stays the same as it was found in the registry.

We would need to see the bootstrapper log and the MSI log from this (in %TEMP%) to see what's wrong.

**Reply by demidov250471** [reply]
Oh, I'm too lazy for that. Let it be as it is. Besides, I'm strongly against repair for esthetic reasons.
Thank you for your support, though.
